lib/puppet/type/cert_key_bundle.rb

```diff
@@ -0,0 +1,96 @@
+Puppet::Type.newtype(:cert_key_bundle) do
+  desc 'combines a certificate and private key into a single file'
+
+  ensurable
+
+  newparam(:path, :namevar => true) do
+    desc "Path to the file that will contain the certificate and private key bundle"
+    isrequired
+  end
+
+  newparam(:certificate) do
+    desc "Path to certificate file to include in bundle"
+    isrequired
+
+    def insync?(is)
+      File.read(self[:path]).include?(File.read(should))
+    end
+  end
+
+  newparam(:private_key) do
+    desc "Path to private key file to include in bundle"
+    isrequired
+
+    def insync?(is)
+      should = format_as_pkcs_1(should) if self[:force_pkcs_1]
+      File.read(self[:path]).include?(File.read(should))
+    end
+  end
+
+  newparam(:owner, parent: Puppet::Type::File::Owner) do
+    desc "Specifies the owner of the file. Valid options: a string containing a username or integer containing a uid."
+  end
+
+  newparam(:group, parent: Puppet::Type::File::Group) do
+    desc "Specifies a permissions group for the file. Valid options: a string containing a group name or integer containing a gid."
+  end
+
+  newparam(:mode, parent: Puppet::Type::File::Mode) do
+    desc "Specifies the permissions mode of the file. Valid options: a string containing a permission mode value in octal notation."
+  end
+
+  newparam(:force_pkcs_1, :boolean => true, :parent => Puppet::Parameter::Boolean) do
+    desc "Ensures the private key is in PKCS#1 format. This is required for services like Apache reverse proxy"
+  end
+
+  def create
+    File.write(self[:path], bundle)
+  end
+
+  def exists?
+    return true if File.exist?(self[:path])
+    bundle == File.read(self[:path])
+  end
+
+  def destroy
+    File.rm(self[:path])
+  end
+
+  def bundle
+    cert = File.read(self[:certificate])
+    key = File.read(self[:private_key])
+
+    key = format_as_pkcs_1(key) if self[:force_pkcs_1]
+
+    [key, cert].join("\n")
+  end
+
+  def format_as_pkcs_1(content)
+    content.gsub(/(BEGIN|END) (PRIVATE KEY)/, '\1 RSA \2')
+  end
```

Comment:
```suggestion
  def bundle
    cert = File.read(self[:certificate])
    if self[:force_pkcs_1]
      key = openssl(
        'rsa',
        '-in', self[:private_key]
    )
    else
      key = File.read(self[:private_key])
    end

    [key, cert].join("\n")
  end
```

Idea I had to implement this with purpose built tools rather than relying on gsub.

This would belong in a provider `lib/puppet/provider/cert_key_bundle/openssl.rb` with `commands openssl: 'openssl'`.

The problem you would face is that Puppet::Provider::Command defaults* to combining STDERR and STDOUT, and overriding this with `options = { combine: false }` drops STDERR completely. Meanwhile openssh sends warning/notice type messages to STDERR, leaving you with unsavory choices of dropping STDERR and not being able to report meaningfully on failure, or parsing the combined STDOUT + STDERR.

*For an even dumber detail, see: https://github.com/puppetlabs/puppet/blob/20efed981cc5f605b98bbb0433a93b457060d5f9/lib/puppet/util/execution.rb#L128-L143

There is an example of working around that with https://github.com/dmitryilyin/puppet-neutron/commit/b55f36a8da62fc207a91b358c396c03c8c58981b which overrides `Puppet::Provider.has_command` with `Open3.capture3` to separate stdout, stderr, and process_status so that stderr can be discarded only when process_status != 0. However I have not tested this.